### ISSUE TITLE
Hello World example

### ISSUE THREAD

**Issue by sudharsanmit** [kubernetes/kops#2340]
I am trying to get k8 on AWS and run the Hello World example of [google](https://cloud.google.com/container-engine/docs/quickstart). 

I understand that it is similar to the [VPC with single public subnet](http://docs.aws.amazon.com/AmazonVPC/latest/UserGuide/VPC_Scenario1.html).

But, kops expects a DNS [configuration](https://github.com/kubernetes/kops/blob/master/docs/aws.md#configure-dns) and recommends not to proceed without this.

As per AWS [documentation](http://docs.aws.amazon.com/AmazonVPC/latest/UserGuide/vpc-dns.html#vpc-dns-hostnames), every instance gets a Public IP and a public DNS by default. Can't this be used as a default, if user doesn't provide one? 

GKE or Azure did not ask me for a DNS configuration for trying this example. Azure needed a DNS prefix while GKE gave me a public IP for testing hello world.

DNS wasn't a constraint when kube-up.sh was used to deploy k8 on AWS as illustrated [here](https://blog.couchbase.com/kubernetes-cluster-amazon-expose-service/).

My kind request - Can you please add a Hello World section which can guide new users to experiment K8 on AWS without having to register a domain and associate it while running the example?

**Comment by neverfox** [comment]
Yeah, coming from a 1.3.5 cluster, this has thrown us a curveball, since we recently stopped using Route 53. I'm sure there's a good reason, but it definitely complicates things, as I cannot even hit the API until I get DevOps to add the entries to GoDaddy (and all the red tape that involves).

**Comment by cheburakshu** [comment]
@neverfox I feel kops is over-engineered and lacks flexibility. I strongly feel, AWS S3 bucket or Route 53 should be preconditions for a cluster creation. Probably they can exist as additional install options. 

I could build a GKE from scratch with terraform, Azure Cluster using az-cli in minutes, without any domain or DNS. AWS - No easy way to onboard.

Not just kops, I had the same problem with CoreOS's kube-aws. A lot of technical documentation, but no easy way to start.

**Comment by cheburakshu** [comment]
Sometimes even comments in kube-up.sh are interesting...

```
# Bring up a Kubernetes cluster.
#
# If the full release name (gs://<bucket>/<release>) is passed in then we take
# that directly.  If not then we assume we are doing development stuff and take
# the defaults in the release config.
```

**Comment by chrislovecnm** [comment]
DNS is being used because of etcd ha lookup. The nodes need to be able to find each other. We are implementing a gossip protocol that will allow for etcd node configuration as well. You are also able to use internal dns with a private zone.

**Comment by neverfox** [comment]
@chrislovecnm Thanks for the reply but my question isn't why DNS is used by k8s per se but rather why, now that kops has replaced kube-up, you're required to have a publically resolving subdomain to even stand a cluster up on AWS? Did something fundamental change in with the architecture? Is there no longer a way to stand up a cluster without the whole domain name process at the beginning?

Before I would simple stand up the cluster and kubectl just worked (because of keys etc), with no thought of domain names. If I needed external services, I used the public DNS of a service ELB. Why the extra hoop to jump through now, and is there a way to get started without a domain name? Are you saying that the private setup is essentially that? Because the examples I've come across (such as [this one](https://www.nivenly.com/kops-1-5-1)) still have the domain step for private setups.

Thanks again for your insight.

**Comment by neverfox** [comment]
#172 at least answers the question about why things have changed.

**Comment by kris-nova** [comment]
So chiming in here. Just wanted to make a few things clear. 

# Support for running without DNS

### I think most of agree that kops should support running without DNS

In fact, we have an open issue for the feature https://github.com/kubernetes/kops/issues/2340

### We use it now, but we want to make it flexible

A lot of the `kops` story is the way it is because we did a lot of things before they were fully baked elsewhere. We (currently) are using Route 53 like @chrislovecnm suggested to look up cluster resources dynamically at runtime. That is **how we use it today** but that doesn't mean **that is how we plan on enforcing it moving forward**. We all want `kops` to be flexible.

### Let's get this onto the roadmap

It seems like this is a hot-button issue and some users have some passion behind it. So let's get it planned and into a future release of `kops`. We release very often. We even have a concept of **feature flags** that allow a user to opt-in to a more alpha feature before we mainstream it. Meaning that we could code in a simple patch for this, and use a **feature flag** to toggle it on/off. Once we push on it and test it a bit more, we can eventually drop the **feature flag** and make it an official feature.

We have a biweekly open call to discuss anything. Anything at all. [More information](https://github.com/kubernetes/kops#office-hours).

We also have a roadmap we could add this feature to. [More information](https://github.com/kubernetes/kops/blob/master/ROADMAP.md) Feel free to open up a PR to get the ball rolling.

### TLDR;

We want to support this, help us help you by helping us plan and code this.

# Hello World guides

Here are some great guides on getting `kops` up and running. I know this outside the original request, but I also have a feeling people will naturally stumble here because of the issue title. So pointers for them:

 - https://github.com/kubernetes/kops/blob/master/docs/aws.md
 - http://deis.com/docs/workflow/quickstart/provider/aws/boot
 - https://www.nivenly.com/kops-1-5-1
 - https://kubernetes.io/docs/getting-started-guides/kops
 - https://daemonza.github.io/2017/01/15/kubernetes-on-aws/
 - https://www.nivenly.com/2nd-hour/


### Closing the issue now because:

###### 1) This is technically a duplicate of https://github.com/kubernetes/kops/issues/2340
###### 2) The hello world will naturally come with the work in https://github.com/kubernetes/kops/issues/2340
###### 3) Also hello world guides are added for anyone else

Is it okay that I close this @neverfox @sudharsanmit @cheburakshu?

Let me know if you still have concerns and we can re-open it.

Cheers everyone

**Comment by cheburakshu** [comment]
@chrislovecnm I have a basic question -
As per K8s literature, etcd is a master component which functions as a backing store for the cluster data. 
Also, again from k8 literature, it is a binding requirement for all k8 clusters to have a cluster DNS.

k8 doesn't mandate DNS for a node. Nowhere in k8s literature it is specified that node should know each other. Node is a worker managed by a master (another node). 

Is kops trying to address a k8s shortcoming for etcd HA?

Point to note is DNS (for etcd HA) is not mandatory on other cloud CLIs (Azure, GKE, etc.).

**Comment by cheburakshu** [comment]
@kris-nova An issue cannot be a duplicate of itself. You have mentioned 

> 1) This is technically a duplicate of #2340

Please re-open or mark the correct dup number.

**Comment by kris-nova** [comment]
@cheburakshu fixed

**Comment by neverfox** [comment]
@kris-nova Thank you for your help and thorough reply.
